src/pal/src/thread/process.cpp

```diff
@@ -3206,153 +3206,247 @@ PROCNotifyProcessShutdown()
+            return false;
         }
     }
     else
     {
         // Gives the child process permission to use /proc/<pid>/mem and ptrace
         if (prctl(PR_SET_PTRACER, childpid, 0, 0, 0) == -1)
         {
-            ERROR("PROCAbort: prctl() FAILED %d (%s)\n", errno, strerror(errno));
+            ERROR("PPROCCreateCrashDump: prctl() FAILED %d (%s)\n", errno, strerror(errno));
+            return false;
         }
         // Parent waits until the child process is done
-        int wstatus;
+        int wstatus = 0;
         int result = waitpid(childpid, &wstatus, 0);
         if (result != childpid)
         {
-            ERROR("PROCAbort: waitpid FAILED result %d wstatus %d errno %d (%s)\n",
+            ERROR("PPROCCreateCrashDump: waitpid FAILED result %d wstatus %d errno %d (%s)\n",
                 result, wstatus, errno, strerror(errno));
+            return false;
         }
+        return !WIFEXITED(wstatus) || WEXITSTATUS(wstatus) == 0;
     }
 #endif // HAVE_PRCTL_H && HAVE_PR_SET_PTRACER
+    return true;
+}
+
+/*++
+Function
+  PROCAbortInitialize()
+
+Abstract
+  Initialize the process abort crash dump program file path and
+  name. Doing all of this ahead of time so nothing is allocated
+  or copied in PROCAbort/signal handler.
+
+Return
+  TRUE - succeeds, FALSE - fails
+
+--*/
+BOOL
+PROCAbortInitialize()
+{
+    char* enabled = getenv("COMPlus_DbgEnableMiniDump");
+    if (enabled != nullptr && _stricmp(enabled, "1") == 0)
+    {
+        char* dumpName = getenv("COMPlus_DbgMiniDumpName");
+        char* dumpType = getenv("COMPlus_DbgMiniDumpType");
+        char* diagStr = getenv("COMPlus_CreateDumpDiagnostics");
+        BOOL diag = diagStr != nullptr && strcmp(diagStr, "1") == 0;
+
+        if (!PROCBuildCreateDumpCommandLine((const char **)g_argvCreateDump, dumpName, dumpType, diag))
+        {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+/*++
+Function:
+  PAL_GenerateCoreDump
+
+Abstract:
+  Public entry point to create a crash dump of the process.
+
+Parameters:
+    dumpName
+    dumpType:
+        Normal = 1,
+        WithHeap = 2,
+        Triage = 3,
+        Full = 4
+    diag
+        true - log createdump diagnostics to console
+
+Return:
+    TRUE success
+    FALSE failed
+--*/
+BOOL
+PAL_GenerateCoreDump(
+    LPCSTR dumpName, 
+    INT dumpType, 
+    BOOL diag)
+{
+    char* argvCreateDump[8] = { nullptr };
+    char dumpTypeStr[16];
+
+    if (dumpType < 1 || dumpType > 4)
+    {
+        return FALSE;
+    }
+    if (dumpName != nullptr && dumpName[0] == '\0')
+    {
+        dumpName = nullptr;
+    }
+    if (_itoa_s(dumpType, dumpTypeStr, sizeof(dumpTypeStr), 10) != 0)
+    {
+        return FALSE;
+    }
+    if (!PROCBuildCreateDumpCommandLine((const char **)argvCreateDump, (char*)dumpName, dumpTypeStr, diag))
+    {
+        return FALSE;
+    }
+    return PROCCreateCrashDump(argvCreateDump);
```

Comment:
@mikem8361 it seems we are leaking the pidarg allocated by the PROCBuildCreateDumpCommandLine